### PR TITLE
ddl: check expr restriction for hash partitioned table

### PR DESCRIPTION
**Pull request by tiancaiamao**
### What problem does this PR solve? 

Mysql:

```
mysql> CREATE TABLE `bb` (`id` int(11) NOT NULL,  
 `ts` TIMESTAMP(6) DEFAULT CURRENT_TIMESTAMP(6), 
 `c` char(120) NOT NULL, 
 KEY `k_ts` (`ts`) ) PARTITION BY   hash(ts);

ERROR 1659 (HY000): Field 'ts' is of a not allowed type for this type of partitioning
```

TiDB:
success

### What is changed and how it works?

The return type of the expr should be int, and only part of functions are allowed
Ref:
https://dev.mysql.com/doc/refman/8.0/en/partitioning-limitations-functions.html
https://dev.mysql.com/doc/refman/8.0/en/partitioning-hash.html

### Check List 

Tests 

 - Unit test

### CODE REVIEW

**Comment by tiancaiamao** [comment]
PTAL @crazycs520

**Comment by tiancaiamao** [comment]
@crazycs520 Hash partition doesn't have "partition by columns ... "

File: ddl/ddl_api.go [hunk 1]
```diff
@@ -1365,41 +1365,41 @@ func buildViewInfoWithTableColumns(ctx sessionctx.Context, s *ast.CreateViewStmt
...
-	if err := checkNoHashPartitions(pi.Num); err != nil {
-		return errors.Trace(err)
+	if err := checkNoHashPartitions(ctx, pi.Num); err != nil {
+		return err
 	}
-	return nil
+	if err := checkPartitionFuncValid(ctx, tbInfo, s.Partition.Expr); err != nil {
+		return err
+	}
+	return checkPartitionFuncType(ctx, s, nil, tbInfo)
```
Comment (crazycs520) [inline ddl/ddl_api.go]:
```suggestion
	return checkPartitionFuncType(ctx, s, tbInfo.Columns, tbInfo)
```
or remove cols paraments, it is redundant?

**Pull request merged**